### PR TITLE
gr-oss-625. Improve retryOperationIfNeeded() output

### PR DESCRIPTION
**Pull request by xBis7**
https://github.com/G-Research/gr-oss/issues/625

This is improving the output of the `retryOperationIfNeeded` method. Each method execution is redirected to a file using the `tee` command. The `tee` command gets the output and prints it to the stdout while also writing it to a file.

I've added an echo with the current command
```
Command being run: -- '...'
```

and I've also wrapped the stdout like this, so that it's easier to spot
```
### Stdout- {

...

} -Stdout ###
```

Example output

```
Command being run: -- 'spark.read.text("hdfs://namenode:8020/test").write.option("path", "hdfs://namenode/opt/hive/data").mode("overwrite").format("csv").saveAsTable("poc_db.spark_test_table")'


### Stdout- {

Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
24/03/29 16:16:37 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://spark-master:4040
Spark context available as 'sc' (master = spark://spark-master:7077, app id = app-20240329161637-0005).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.3.2
      /_/
         
Using Scala version 2.12.15 (OpenJDK 64-Bit Server VM, Java 11.0.22)
Type in expressions to have them evaluated.
Type :help for more information.

scala> Hive Session ID = 582b5268-b4f7-4c27-8618-6eed1dbd59a1
24/03/29 16:16:44 WARN HiveExternalCatalog: Couldn't find corresponding Hive SerDe for data source provider csv. Persisting data source table `poc_db`.`spark_test_table` into Hive metastore in Spark SQL specific format, which is NOT compatible with Hive.
org.apache.spark.sql.AnalysisException: org.apache.hadoop.hive.ql.metadata.HiveException: MetaException(message:Permission denied: user [spark] does not have [CREATE] privilege on [poc_db/spark_test_table])
  at org.apache.spark.sql.hive.HiveExternalCatalog.withClient(HiveExternalCatalog.scala:110)
...
...
...
...
...
...
...
...
...
  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
  at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
  at java.base/java.lang.reflect.Method.invoke(Unknown Source)
  at org.apache.hadoop.hive.metastore.RetryingMetaStoreClient.invoke(RetryingMetaStoreClient.java:212)
  at com.sun.proxy.$Proxy40.createTable(Unknown Source)
  at org.apache.hadoop.hive.ql.metadata.Hive.createTable(Hive.java:928)
  ... 93 more

scala> :quit


} -Stdout ###
```

I've run all tests for both Spark and Trino multiple times to make sure nothing is broken.

The tricky part has been printing the command. If we add the print inside every method that runs the command, then the print will also get stored in the tmp file that we are checking for the expected output message. Some times the expected output is part of the command and we end up with unexpected failures.

### CODE REVIEW

**Review by mladjan-gadzic — APPROVED** [review]
Handy improvement, thanks! I've tested everything - no failures.

File: trino-hms-hdfs-ranger/testlib.sh [hunk 1]
```diff
@@ -565,142 +570,222 @@ updateProjectRepo() {
...
   hdfs_dir_name=$2
   db_name=$3
+  print=$4
 
-  if [ "$db_name" == "" ]; then
-    db_name="default"
+  if [ "$print" == "true" ]; then
+    printCmdString "spark.read.text(\"hdfs://namenode:8020/$hdfs_dir_name\").write.option(\"path\", \"hdfs://namenode/opt/hive/data\").mode(\"overwrite\").format(\"csv\").saveAsTable(\"$db_name.$table_name\")"
+  else
+    docker exec -it "$SPARK_MASTER_HOSTNAME" bash -c "echo \"spark.read.text(\\\"hdfs://namenode:8020/$hdfs_dir_name\\\").write.option(\\\"path\\\", \\\"hdfs://namenode/opt/hive/data\\\").mode(\\\"overwrite\\\").format(\\\"csv\\\").saveAsTable(\\\"$db_name.$table_name\\\")\" | bin/spark-shell"
```
Comment (GeorgeJahad) [inline trino-hms-hdfs-ranger/testlib.sh:639]: I don't like how we have two differing copies of the command in each of these subcommands.  I would much prefer a structure like this:

```
cmd = xyz
if print == true
   printCmdString $cmd
else $cmd
```

File: trino-hms-hdfs-ranger/testlib.sh [hunk 2]
```diff
@@ -839,13 +924,45 @@ retryOperationIfNeeded() {
...
+    # * The expected output is just the phrase 'CREATE SCHEMA'
+    # * The command to create a schema, starts with the phrase 'CREATE SCHEMA ...'
+    # * We do a 'grep' on the tmp file to check that the output contains 'CREATE SCHEMA'
+    # * The command failed and the output is some error
+    # * 'CREATE SCHEMA' exists in the command, which is present in the tmp file.
+    # * 'grep' succeeds, exit status will be 0 regardless of the output.
+    #
+    # To fix that, each method will either print the command or execute it.
+    # We will call each method twice. Once with the print parameter and once without.
+    $cmd "true"
```
Comment (GeorgeJahad) [inline trino-hms-hdfs-ranger/testlib.sh:942]: It is not great to have the last parameter have a special meaning.  Now people have to remember when they add a parameter to make it the second to last.  It is not very elegant.  Can you do something else?  Maybe use a global variable?

Comment (xBis7) [reply trino-hms-hdfs-ranger/testlib.sh]: Done.

File: trino-hms-hdfs-ranger/testlib.sh (hunk 1 above)
Comment (xBis7) [reply trino-hms-hdfs-ranger/testlib.sh:639]: That was my first approach but it's a bit tricky to properly handle the quotation marks. I'll try to find a way.

**Comment by xBis7** [comment]
@mladjan-gadzic @GeorgeJahad Can you take another look? I've made all the changes and I've also run all tests for Spark and Trino and they all pass.

File: trino-hms-hdfs-ranger/testlib.sh (hunk 1 above)
Comment (xBis7) [reply trino-hms-hdfs-ranger/testlib.sh]: Fixed.

**Review by GeorgeJahad — APPROVED** [review]
Thanks for the fixes.  LGTM.  Merging so that Mladjan can integrate his fixes.

**Pull request merged**